src/main/java/edu/cornell/kfs/sys/cache/ClientTrackingLettuceConnectionFactory.java

```diff
@@ -0,0 +1,67 @@
+package edu.cornell.kfs.sys.cache;
+
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.springframework.beans.factory.DisposableBean;
+import org.springframework.data.redis.connection.RedisStandaloneConfiguration;
+import org.springframework.data.redis.connection.lettuce.LettuceClientConfiguration;
+import org.springframework.data.redis.connection.lettuce.LettuceConnectionFactory;
+import org.springframework.data.redis.connection.lettuce.LettuceConnectionProvider;
+
+import io.lettuce.core.AbstractRedisClient;
+import io.lettuce.core.codec.RedisCodec;
+
+public class ClientTrackingLettuceConnectionFactory extends LettuceConnectionFactory {
+
+    private static final Logger LOG = LogManager.getLogger();
+
+    private final RedisEventListener eventListener;
+    private final AtomicBoolean initializedProviderForTracking;
+
+    public ClientTrackingLettuceConnectionFactory(RedisStandaloneConfiguration standaloneConfig,
+            LettuceClientConfiguration clientConfig, RedisEventListener eventListener) {
+        super(standaloneConfig, clientConfig);
+        Objects.requireNonNull(eventListener, "eventListener cannot be null");
+        this.eventListener = eventListener;
+        this.initializedProviderForTracking = new AtomicBoolean(false);
+    }
+
+    @Override
+    protected LettuceConnectionProvider doCreateConnectionProvider(AbstractRedisClient client,
+            RedisCodec<?, ?> codec) {
+        LettuceConnectionProvider connectionProvider = super.doCreateConnectionProvider(client, codec);
+        if (initializedProviderForTracking.compareAndSet(false, true)) {
```

Comment:
I am not sure I understand how this works.  initializedProviderForTracking is defaulted to false, so compareAndSet would be true, because we define the expected value is false, and then set the value to true.  However, initializedProviderForTracking is also final, so that means the value should not be changeable.

How does the else condition get executued?

Reply:
initializedProviderForTracking is an AtomicBoolean, which is a wrapper object and not a primitive boolean. The declaration of "final" only makes the reference to the wrapper object immutable; it does not make its internal boolean value immutable.

The superclass will call the doCreateConnectionProvider() method at least twice: Once to initialize a regular shared connection, and a second time to initialize a shared connection to be used with the "reactive" API (which this PR does not use at all). The setup of the "reactive" connection should get handled by the "else" block, and the DEBUG message in this code explains why we don't want client tracking on both connections.

Reply:
Thanks for clarification